### DISCUSSION TITLE
ECMR algorithm for encryption in browser

### DISCUSSION THREAD

**Discussion by clauderobi** [panva/jose#668] [Q&A]
I am using tang / clevis and able to replicate the the encryption operation of clevis using this jose library. But only in a node JS context. My need is actually for a browser context where the ECMR is not recognized (or implemented).

I suspect it is possible to perform the ECMR procedure manually and I tried bit the resulting JWE does not work with the desktop clevis client. So I am doing something wrong. I simply specified ECDH-ES for the encryption key and replaced alg back to ECMR in the returned JWE (was not expecting it to work though...). Any suggestion?

**Comment by panva** [comment]
`ECMR` is not a recognized Algorithm Identifier in any runtime, node or browser. It is not [registered](https://www.iana.org/assignments/jose/jose.xhtml) and therefore as far as i'm concerned, non existant.

If there are tools with proprietary algorithms then use those.

**Reply by clauderobi** [reply]
Under nodeJS ECMR is indeed available since I successfully used it. It is also supported by the C variant of Jose. Because of the name (and timeline) I was under the impression the module here was a reimplementation of the C Jose but it is not the case.

So, WASM is the way to go.

**Reply by panva** [reply]
I cannot speak to other unrelated projects that choose to implement proprietary and unregistered JOSE algorithms.

**Reply by clauderobi** [reply]
I understand, except there are 2 things:
- That unrelated project is actually jose (the C one) and nodeJS is using it since ECMR works with your project under nodeJS
- ECMR was indeed included in other libraries that nodeJS uses, such as openSSL.

**Reply by panva** [reply]
ECMR is not a recognized algorithm by this very project.

**Reply by clauderobi** [reply]
Interesting,...
but yet using ECMR with the nodeJS runtime worked.

Scary.

**Reply by panva** [reply]
> but yet using ECMR with the nodeJS runtime worked.

Fairly certain it didn't. If the identifier is not recognized, it cannot be neither produced nor consumed by tis project.

Please provide an example.

**Reply by clauderobi** [reply]
What you describes is exactly what happens when running in a browser, which is my real target. For dev purpose I ran it in nodeJs first and successfully created a working JWE (using tang as a public key provider for the purpose of key exchange). 

After moving to testing with a browser and seeing that it did not work I played with ECDH-ES instead but that broke the tang protocol and then did a few other attempts. In the end I gave up and went the wasm way. Unfortunately, I did not keep the code that was working under nodeJS. If you really really want it I could recreate.

I have a question; who is not recognizing the identifier? This project's code or something in the run time environment?

**Reply by panva** [reply]
> who is not recognizing the identifier? This project's code or something in the run time environment?

The project. I only strive to implement registered identifiers.

> Unfortunately, I did not keep the code that was working under nodeJS. If you really really want it I could recreate.

Please. I can't see a way that you would be able to use the `ECMR` identifier using this project's module in any runtime.
